Work mzmt admin logic

MaximeZmt opened the pull request:
Thank you for opening a pull request with this project, please also:

* [ ] assign the PR to yourself, or to the person(s) working on it
* [ ] add a brief description of your changes here
* [ ] start in `draft` mode and `in progress` pipeline in the project (if applicable)
* [ ] if applicable, add this PR to its related issue by one of the special keywords [Closing keywords](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue)
* once it's ready 
  * [ ] put it in the `Review` or `Ready4Merge` pipeline in the project (if applicable) 
  * [ ] remove `draft`
  * [ ] assign a reviewer

Commit:
start implement is admin logic

Commit:
ongoing implem

Commit:
patch changes

Commit:
patch changes 2

Commit:
Init AdminForm method

Commit:
now need an admin perm on blockchain to create form

MaximeZmt committed:
Merge branch 'main' into Work-Mzmt-AdminLogic

Commit:
evening work refactor part 1

Commit:
merge rebase

Commit:
patch one test

Commit:
Merge branch 'Work-Mzmt-Admin-Logic-Patch' into Work-Mzmt-AdminLogic

Commit:
patch vet

Commit:
patch even more adminform to adminlist

Commit:
getting rid of adminform

Commit:
add performing and target field

Commit:
patch error and test

Commit:
some minor changes

Commit:
small code cleaning

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -783,36 +797,79 @@ func (e evotingCommand) deleteForm(snap store.Snapshot, step execution.Step) err
+		form, err = e.getAdminList(snap)
 		if err != nil {
-			return xerrors.Errorf("failed to get AdminForm: %v", err)
+			if !strings.Contains(err.Error(), "No form found") {
```
is it possible to match the exact error string here? if someone else would create an error message like "No form found because of some other error" it would not match despite the list possibly existing

that any error message starting by "No form found" will lead to the code trying to create the chain is implicit knowledge that makes it hard to debug

alternatively you could use a custom error

MaximeZmt replied:
I've pushed a new version with a matching on the exact error string.

Then as it is with a Not in front, basically as soon as it contains "No form found" it blocks in a fail safe by avoiding it.

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -783,36 +797,79 @@ func (e evotingCommand) deleteForm(snap store.Snapshot, step execution.Step) err
+		// Check that the performing user is Admin
+		performingUserPerm, err := form.GetAdminIndex(txAddAdmin.PerformingUserID)
+		if err != nil {
+			return xerrors.Errorf("couldn't retrieve admin permission of the performing user: %v", err)
```
maybe add the SCIPER to the error here as well

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -783,36 +797,79 @@ func (e evotingCommand) deleteForm(snap store.Snapshot, step execution.Step) err
+
+		err = form.AddAdmin(txAddAdmin.TargetUserID)
 		if err != nil {
 			return xerrors.Errorf("couldn't add admin: %v", err)
```
maybe add the SCIPER of the admin if it isn't included in the error message

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -783,36 +797,79 @@ func (e evotingCommand) deleteForm(snap store.Snapshot, step execution.Step) err
-		err = form.AddAdmin(txAddAdmin.UserID)
+		// If it found the AdminList
+		// Check that the performing user is Admin
+		performingUserPerm, err := form.GetAdminIndex(txAddAdmin.PerformingUserID)
+		if err != nil {
+			return xerrors.Errorf("couldn't retrieve admin permission of the performing user: %v", err)
+		}
+
+		if performingUserPerm < 0 {
+			return xerrors.Errorf("the performing user %v doesn't have the permission to add admin", txAddAdmin.PerformingUserID)
+		}
+
```
maybe you could move this to a function `is_admin`

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -825,7 +882,64 @@ func (e evotingCommand) manageAdminForm(snap store.Snapshot, step execution.Step
+
+	formBuf, err := adminList.Serialize(ctx)
+	if err != nil {
+		return xerrors.Errorf("failed to marshal Admin Form : %v", err)
```
```suggestion
		return xerrors.Errorf("failed to marshal AdminList : %v", err)
```

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -825,7 +882,64 @@ func (e evotingCommand) manageAdminForm(snap store.Snapshot, step execution.Step
+		return xerrors.Errorf("failed to set value: %v", err)
+	}
+
+	// Update the form metadata store
+
+	formsMetadataBuf, err := snap.Get([]byte(FormsMetadataKey))
+	if err != nil {
+		return xerrors.Errorf("failed to get key '%s': %v", formsMetadataBuf, err)
+	}
+
+	formsMetadata := &types.FormsMetadata{
+		FormsIDs: types.FormIDs{},
+	}
+
+	if len(formsMetadataBuf) != 0 {
+		err := json.Unmarshal(formsMetadataBuf, formsMetadata)
+		if err != nil {
+			return xerrors.Errorf("failed to unmarshal FormsMetadata: %v", err)
+		}
+	}
+
+	err = formsMetadata.FormsIDs.Add(hex.EncodeToString(formIDBuf))
+	if err != nil {
+		return xerrors.Errorf("couldn't add new form: %v", err)
+	}
+
+	formMetadataJSON, err := json.Marshal(formsMetadata)
+	if err != nil {
+		return xerrors.Errorf("failed to marshal FormsMetadata: %v", err)
+	}
+
+	err = snap.Set([]byte(FormsMetadataKey), formMetadataJSON)
```
this seems to me to be code that's probably used elsewhere too - if that's the case, maybe put it in a function that's called?

PascalinDe commented on contracts/evoting/mod_test.go:
```diff
@@ -1109,13 +1110,21 @@ func TestRegisterContract(t *testing.T) {
+		- Initialize Contract & Form
+		- Perform error handling test
+		- 123456 add admin 123456
+		- 123456 remove admin 123456 -> error cause only one left
+		- 777777 add admin 777777 -> error cause not an admin
```
```suggestion
		- 123456 remove admin 123456 -> error because only one left
		- 777777 add admin 777777 -> error because not an admin
```

PascalinDe commented on contracts/evoting/types/admin.go:
```diff
@@ -45,26 +42,26 @@ func (adminForm AdminForm) Deserialize(ctx serde.Context, data []byte) (serde.Me
 	sciperInt, err := SciperToInt(userID)
 	if err != nil {
-		return xerrors.Errorf("failed to convert SCIPER to integer: %v", err)
+		return xerrors.Errorf("Failed SciperToInt: %v", err)
```
```suggestion
		return xerrors.Errorf("Failed to convert SCIPER to int: %v", err)
```

PascalinDe replied:
did you change it back by accident?

PascalinDe commented on contracts/evoting/types/admin.go:
```diff
@@ -45,26 +42,26 @@ func (adminForm AdminForm) Deserialize(ctx serde.Context, data []byte) (serde.Me
 	sciperInt, err := SciperToInt(userID)
 	if err != nil {
-		return -1, xerrors.Errorf("failed to convert SCIPER to integer: %v", err)
+		return -1, xerrors.Errorf("Failed SciperToInt: %v", err)
```
```suggestion
		return -1, xerrors.Errorf("Failed to convert SCIPER to int: %v", err)
```

PascalinDe replied:
same

PascalinDe commented on contracts/evoting/types/admin.go:
```diff
@@ -73,58 +70,63 @@ func (adminForm *AdminForm) GetAdminIndex(userID string) (int, error) {
+	index, err := adminList.GetAdminIndex(userID)
 	if err != nil {
-		return xerrors.Errorf("Failed to retrieve the Admin permission: %v", err)
+		return xerrors.Errorf("Failed GetAdminIndex: %v", err)
 	}
 
 	if index < 0 {
 		return xerrors.Errorf("Error while retrieving the index of the element.")
 	}
 
```
please change this to more intuitive error messages (if I don't know the code very well, I don't know what is `GetAdminIndex` or what it means that the index could not be found)

PascalinDe replied:
also changed back

PascalinDe commented on contracts/evoting/types/admin.go:
```diff
@@ -73,58 +70,63 @@ func (adminForm *AdminForm) GetAdminIndex(userID string) (int, error) {
 	}
 
-	adminForm.AdminList = append(adminForm.AdminList[:index], adminForm.AdminList[index+1:]...)
+	// We don't want to have a form without any Owners.
```
this comment does not seem to fit here?

MaximeZmt committed:
Update contracts/evoting/evoting.go

Co-authored-by: Carine Dengler <github@pascalin.de>

MaximeZmt committed:
Update contracts/evoting/mod_test.go

Co-authored-by: Carine Dengler <github@pascalin.de>

MaximeZmt committed:
Update contracts/evoting/types/admin.go

Co-authored-by: Carine Dengler <github@pascalin.de>

MaximeZmt committed:
Update contracts/evoting/types/admin.go

Co-authored-by: Carine Dengler <github@pascalin.de>

Commit:
address partial review

Commit:
isAdmin function

Commit:
patch updatemetadata review

Commit:
patch ownervoter test

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -750,69 +771,77 @@ func (e evotingCommand) deleteForm(snap store.Snapshot, step execution.Step) err
+		form, err = e.getAdminList(snap)
 		if err != nil {
-			return xerrors.Errorf("failed to get AdminForm: %v", err)
+			// Exact string matching of the error
+			if !strings.Contains(err.Error(), "failed to get the AdminList: No form found") {
```
very minor remark, if it's an exact match now, can't you use the comparison operator?

MaximeZmt replied:
Yes, will do the update 👍

MaximeZmt committed:
Update evoting.go

exact error comparison update

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -65,6 +65,20 @@ func (e evotingCommand) createForm(snap store.Snapshot, step execution.Step) err
 	}
 
+	// Check if has Admin Right to create a form
+	adminList, err := e.getAdminList(snap)
+	if err != nil {
+		return xerrors.Errorf("failed to get AdminList: %v", err)
+	}
+	adminIndex, err := adminList.GetAdminIndex(tx.UserID)
+	if err != nil {
+		return xerrors.Errorf("Couldn't retrieve the admin right of the user: %v", err)
+	}
+
+	if adminIndex < 0 {
+		return xerrors.Errorf("The user is not admin: %v", err)
+	}
```
can this be part of the `is_admin` function?

MaximeZmt replied:
My problem by moving this part to the is_admin is that it is used to check whether the Admin List already exist and in case it does not, create the Admin List (initialize admin list procedure) -> Therefore I would find weird that it is the is_admin that perform the logic of initializing the Admin List.

What is your opinion about it ?

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -825,14 +854,58 @@ func (e evotingCommand) manageAdminForm(snap store.Snapshot, step execution.Step
+	}
+
+	if performingUserPerm < 0 {
+		return false, xerrors.Errorf("the performing user %v doesn't have the permission to add admin", txPerformingUser)
+	}
```
the function is behaving correctly (returning false if the admin cannot be found), so there's no error

MaximeZmt committed:
patch is admin error mgmt

MaximeZmt committed:
patch test

MaximeZmt committed:
patch isAdmin structure

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -820,42 +816,38 @@ func (e evotingCommand) manageAdminList(snap store.Snapshot, step execution.Step
+		list, err = e.getAdminList(snap)
 		if err != nil {
 			return xerrors.Errorf("failed to get AdminList: %v", err)
 		}
```
with the changes to `is_admin` this part of the code is no longer necessary (as this will be checked in `is_admin` too)

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -869,18 +861,23 @@ func (e evotingCommand) manageAdminList(snap store.Snapshot, step execution.Step
+	if err != nil {
+		return false, types.AdminList{}, err
+	}
+
```
I would add the error checking here (if list cannot be retrieved because of an error) instead of doing it seperately in the calling functions, since retrieving the list successfully is part of this functions functionality now

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -869,18 +861,23 @@ func (e evotingCommand) manageAdminList(snap store.Snapshot, step execution.Step
 
 // isAdmin Check whether a user is in an Admin List
-func (e evotingCommand) isAdmin(form types.AdminList, txPerformingUser string) (bool, error) {
+func (e evotingCommand) isAdmin(snap store.Snapshot, txPerformingUser string) (bool, types.AdminList, error) {
```
maybe `fetchAdmin` or something similar would be a more appropriate name now

PascalinDe commented on contracts/evoting/evoting.go:
```diff
@@ -796,11 +791,12 @@ func (e evotingCommand) manageAdminList(snap store.Snapshot, step execution.Step
 		if err != nil {
 			// Exact string matching of the error
-			if err.Error() != "failed to get the AdminList: No form found" {
+			println("HEYO: " + err.Error())
```
please remove this debugging print

Commit:
Address review
fixed comment
change isAdmin to fetchAdmin
remove useless getAdminList in manageAdminList cmd
